File: plsql-checks/src/main/java/br/com/felipezorzo/sonar/plsql/checks/scope/Scope.java
/*
 * Sonar PL/SQL Plugin (Community)
 * Copyright (C) 2015 Felipe Zorzo
 * felipe.b.zorzo@gmail.com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02
 */
package br.com.felipezorzo.sonar.plsql.checks.scope;

import java.util.HashMap;
import java.util.Map;

import javax.annotation.Nullable;

import com.google.common.base.Preconditions;
import com.sonar.sslr.api.AstNode;

import br.com.felipezorzo.sonar.plsql.api.PlSqlGrammar;

public class Scope {

    private Map<String, Variable> localVariables = new HashMap<>();
    private Scope outerScope;
    
    public Scope(Scope outerScope) {
        this.outerScope = outerScope;
    }
    
    public void declareLocalVariable(AstNode node) {
        declareLocalVariable(node, 0);
    }
    
    public void declareLocalVariable(AstNode node, int usage) {
        Preconditions.checkArgument(node.is(PlSqlGrammar.IDENTIFIER_NAME));
        localVariables.put(node.getTokenOriginalValue().toUpperCase(), new Variable(node, usage));
    }

    public Map<String, Variable> getLocalVariables() {
        return localVariables;
    }
    
    @Nullable
    public Variable getVariableDeclaration(AstNode node) {
        String variableName = node.getTokenOriginalValue().toUpperCase();
        
        Scope scope = this;
        while (scope != null) {
            Map<String, Variable> variables = scope.getLocalVariables();
            if (variables.containsKey(variableName)) {
                return variables.get(variableName);
            }
            scope = scope.outerScope;
        }
        return null;
    }

    public void useVariable(AstNode node) {
        Variable variable = getVariableDeclaration(node);
        if (variable != null) {
            variable.increaseUsage();
        }
    }

}

File: zpa-core/src/main/kotlin/com/felipebz/zpa/api/symbols/Scope.kt
/**
 * Z PL/SQL Analyzer
 * Copyright (C) 2015-2025 Felipe Zorzo
 * mailto:felipe AT felipezorzo DOT com DOT br
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package com.felipebz.zpa.api.symbols

import com.felipebz.flr.api.AstNode
import com.felipebz.flr.api.AstNodeType
import com.felipebz.flr.api.Token
import com.felipebz.zpa.api.PlSqlFile
import java.util.*

interface Scope {
    val isAutonomousTransaction: Boolean
    val isOverridingMember: Boolean
    val symbols: List<Symbol>
    val tree: AstNode?
    val outer: Scope?
    val identifier: String?
    val hasExceptionHandler: Boolean
    val path: List<String>
    val innerScopes: List<Scope>
    val type: AstNodeType?
    val isGlobal: Boolean
    val plSqlFile: PlSqlFile?
    val firstToken: Token?
    val lastToken: Token?

    /**
     * @param kind of the symbols to look for
     * @return the symbols corresponding to the given kind
     */
    fun getSymbols(kind: Symbol.Kind): List<Symbol>

    fun getSymbolsAcessibleInScope(name: String, vararg kinds: Symbol.Kind): Deque<Symbol>
    fun addSymbol(symbol: Symbol)
    fun getSymbol(name: String, vararg kinds: Symbol.Kind): Symbol?
    fun getSymbol(name: String, path: List<String>, vararg kinds: Symbol.Kind): Symbol?
    fun addInnerScope(scope: Scope)
}
